lsp-mode.el

```diff
@@ -3905,6 +3916,178 @@ A reference is highlighted only if it is visible in a window."
+  "Face used for semantic highlighting scopes matching variable.other.field.static.*,
+unless overriden by a more specific face association."
+  :group 'lsp-faces)
+
+(defface lsp-face-semhl-enummember
+  '((t :inherit font-lock-constant-face))
+  "Face used for semantic highlighting scopes matching variable.other.enummember.*,
+unless overriden by a more specific face association."
+  :group 'lsp-faces)
+
+(defface lsp-face-semhl-variable
+  '((t :inherit font-lock-variable-name-face))
+  "Face used for semantic highlighting scopes matching variable.*,
+unless overriden by a more specific face association."
+  :group 'lsp-faces)
+
+(defface lsp-face-semhl-function
+  '((t :inherit font-lock-function-name-face))
+  "Face used for semantic highlighting scopes matching entity.name.function.*,
+unless overriden by a more specific face association."
+  :group 'lsp-faces)
+
+(defface lsp-face-semhl-method
+  '((t :inherit lsp-face-semhl-function))
+  "Face used for semantic highlighting scopes matching entity.name.function.method.*,
+unless overriden by a more specific face association."
+  :group 'lsp-faces)
+
+(defface lsp-face-semhl-static-method
+  '((t :inherit lsp-face-semhl-function :italic ))
+  "Face used for semantic highlighting scopes matching entity.name.function.method.static.*,
+unless overriden by a more specific face association."
+  :group 'lsp-faces)
+
+(defface lsp-face-semhl-type-class
+  '((t :inherit font-lock-type-face))
+  "Face used for semantic highlighting scopes matching entity.name.type.class.*,
+unless overriden by a more specific face association."
+  :group 'lsp-faces)
+
+(defface lsp-face-semhl-type-enum
+  '((t :inherit font-lock-type-face))
+  "Face used for semantic highlighting scopes matching entity.name.type.enum.*,
+unless overriden by a more specific face association."
+  :group 'lsp-faces)
+
+(defface lsp-face-semhl-namespace
+  '((t :inherit font-lock-type-face :bold t))
+  "Face used for semantic highlighting scopes matching entity.name.namespace.*,
+unless overriden by a more specific face association."
+  :group 'lsp-faces)
+
+(defface lsp-face-semhl-preprocessor
+  '((t :inherit font-lock-preprocessor-face))
+  "Face used for semantic highlighting scopes matching entity.name.function.preprocessor.*,
+unless overriden by a more specific face association."
+  :group 'lsp-faces)
+
+(defface lsp-face-semhl-type-template
+  '((t :inherit font-lock-type-face :italic t))
+  "Face used for semantic highlighting scopes matching entity.name.type.template.*,
+unless overriden by a more specific face association."
+  :group 'lsp-faces)
+
+(defface lsp-face-semhl-type-primitive
+  '((t :inherit font-lock-type-face :italic t))
+  "Face used for semantic highlighting scopes matching storage.type.primitive.*,
+unless overriden by a more specific face association."
+  :group 'lsp-faces)
+
+(defvar lsp-semantic-highlighting-faces
+  '(("^variable\\.parameter\\(\\..*\\)?$" . lsp-face-semhl-variable-parameter)
```

Comment:
Is there something that forces us to use regex but not prefix and then test with starts-with? My main concern is that the regexp will be much slower.

Reply:
regexp matching is only done on the first fontification run (within the `(unless lsp--facemap...` part in `lsp--on-semantic-highlighting`), and only needs to traverse the `scopes` vector sent by the language server (may have a dozen members or so).
For the fontification itself, faces are retrieved directly from `lsp--facemap` using `scopeIndex`

Reply:
Fair enough.